File: lc-python/src/lc_0014_longest_common_prefix.py
import unittest
from typing import List


def longest_common_prefix0(ws: List[str]) -> str:
    """
    Time complexity: O(kn)
    Space complexity: O(1)

    where
    - n is the number of words, and
    - k is the length of the shortest word
    """
    if len(ws) == 0:
        return ""

    # length of shortest word
    l = min(map(len, ws))

    for k in range(l):
        ch = ws[0][k]
        for i in range(1, len(ws)):
            w = ws[i]
            if ch != w[k]:
                return ws[0][:k]

    # occurs when ws has only 1 element or
    # when all elements are the empty string
    return ws[0][:l]


def longest_common_prefix(ws: List[str]) -> str:
    """
    Time complexity: O(kn log n)
    - The sort is O(n log n) but each comparison in the sort is O(k)...
    Space complexity: O(kn)

    where
    - n is the number of words, and
    - k is the length of the shortest word
    """
    if len(ws) == 0:
        return ""

    if len(ws) == 1:
        return ws[0]

    # sort string alphabetically
    ws = sorted(ws)
    word1 = ws[0]
    word2 = ws[-1]

    # compare only the first and last words
    ans = ""
    l = min(len(word1), len(word2))
    for i in range(l):
        if word1[i] != word2[i]:
            break
        ans += word1[i]

    return ans


class TestLongestCommonPrefix(unittest.TestCase):
    @staticmethod
    def parse_input():
        data = [
            [[], ""],
            [[""], ""],
            [["pewpew"], "pewpew"],
            [["", ""], ""],
            [["flower", "flow", "flight"], "fl"],
            [["dog", "racecar", "car"], ""],
            [["cir", "car"], "c"],
            [["reflower", "flow", "flight"], ""],
        ]
        return data

    def test_longest_common_prefix0(self):
        test_cases = self.parse_input()
        for tc in test_cases:
            ws = tc[0]
            expected = tc[1]

            actual = longest_common_prefix0(ws)

            self.assertEqual(actual, expected)

    def test_longest_common_prefix(self):
        test_cases = self.parse_input()
        for tc in test_cases:
            ws = tc[0]
            expected = tc[1]

            actual = longest_common_prefix(ws)

            self.assertEqual(actual, expected)


if __name__ == "__main__":
    unittest.main()

File: lc-typescript/src/lc_0014_longest_common_prefix.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/assert_equals.ts";

function longest_common_prefix(strs: string[]): string {
  if (strs.length <= 1) {
    return strs[0] || "";
  }

  const j_max = strs.reduce((acc, curr) => Math.max(acc, curr.length), 0);

  let prefix = "";
  let mismatch = false;
  let j = 0;
  while (j < j_max) {
    const ch = strs[0][j];
    for (let i = 1; i < strs.length; i += 1) {
      if (ch != strs[i][j]) {
        mismatch = true;
        break;
      }
    }

    if (mismatch) {
      break;
    }

    prefix += ch;
    j += 1;
  }

  return prefix;
}

Deno.test("0014 Longest Common Prefix", () => {
  const test_cases: [string[], string][] = [
    [[], ""],
    [[""], ""],
    [["pewpew"], "pewpew"],
    [["", ""], ""],
    [["flower", "flow", "flight"], "fl"],
    [["dog", "racecar", "car"], ""],
    [["cir", "car"], "c"],
    [["reflower", "flow", "flight"], ""],
  ];

  for (let i = 0; i < test_cases.length; i += 1) {
    const test_case = test_cases[i];

    const strs = test_case[0];
    const expected = test_case[1];

    const actual = longest_common_prefix(strs);
    assertEquals(actual, expected);
  }
});
